aead/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+    /// The maximum length of the nonce.
+    type TagSize: ArrayLength<u8> + Unsigned;
+    /// The amount of suffix padding, in bytes, which needs to be
+    /// appended to a plaintext to accomodate this cipher's output.
+    type CiphertextOverhead: ArrayLength<u8> + Unsigned;
+
+    /// Encrypts the given plaintext into a new ciphertext object and the nonce
+    fn encrypt<'in_out, AdItem: AsRef<[u8]>, AdIter: Iterator<Item = AdItem>>(
+        key: &GenericArray<u8, Self::KeySize>,
+        additional_data: AdIter,
+        nonce: &mut GenericArray<u8, Self::NonceSize>,
+        plaintext: &'in_out mut [u8],
+        plaintext_used: usize,
+    ) -> Result<&'in_out mut [u8], Error>;
+
+    /// Authenticates the ciphertext, nonce, and additional data, then
+    /// decrypts the ciphertext contents into plaintext.
+    fn decrypt<'in_out, AdItem: AsRef<[u8]>, AdIter: Iterator<Item = AdItem>>(
+        key: &GenericArray<u8, Self::KeySize>,
+        additional_data: AdIter,
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        ciphertext: &'in_out mut [u8],
+    ) -> Result<&'in_out mut [u8], Error>;
+}
+
+/// A wrapper structure to allow use of a stateless AEAD through the stateful
+/// interface.
+pub struct Stateful<Algo: StatelessAead> {
+    key: GenericArray<u8, Algo::KeySize>,
+    _aead: PhantomData<fn() -> Algo>
+}
+
+impl<Algo: StatelessAead> Aead for Stateful<Algo> {
+    type KeySize = Algo::KeySize;
+    type NonceSize = Algo::NonceSize;
+    type TagSize = Algo::TagSize;
+    type CiphertextOverhead = Algo::CiphertextOverhead;
+
+    fn for_key(key: GenericArray<u8, Self::KeySize>) -> Self {
+        Self {
+            key,
+            _aead: PhantomData::default(),
+        }
+    }
+
+    fn encrypt<'in_out, AdItem: AsRef<[u8]>, AdIter: Iterator<Item = AdItem>>(
+        &mut self,
+        additional_data: AdIter,
+        nonce: &mut GenericArray<u8, Self::NonceSize>,
+        plaintext: &'in_out mut [u8],
+        plaintext_used: usize,
+    ) -> Result<&'in_out mut [u8], Error> {
+        Algo::encrypt(&self.key, additional_data, nonce, plaintext, plaintext_used)
+    }
+
+    fn decrypt<'in_out, AdItem: AsRef<[u8]>, AdIter: Iterator<Item = AdItem>>(
+        &mut self,
+        additional_data: AdIter,
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        ciphertext: &'in_out mut [u8],
+    ) -> Result<&'in_out mut [u8], Error> {
+        Algo::decrypt(&self.key, additional_data, nonce, ciphertext)
+    }
+}
+
+#[cfg(any(feature = "alloc", all(feature = "std", has_extern_crate_alloc)))]
+extern crate alloc;
+
+#[cfg(any(feature = "alloc", all(feature = "std", has_extern_crate_alloc)))]
+use alloc::vec::Vec;
+#[cfg(all(feature = "std", not(any(feature = "alloc", has_extern_crate_alloc))))]
+use std::vec::Vec;
+
+#[cfg(any(feature = "alloc", feature = "std"))]
+/// Users who wish to use vectors instead of mutable byte slices should
+/// utilize this API.
+pub trait AeadVec<Algo: Aead> {
```

Comment:
I thought the same thing, but there was conversation previously around doing this as an extension trait to prevent authors from fiddling with stuff. I'm fine with adding it back to the trait itself, that leaves open the opportunity for algos to get slick.

Reply:
Well, to use those methods you will have to activate `alloc` feature, so I don't see how it could be a problem. Plus we already have such convenience methods for example in `BlockMode`,  `digest::VariableOutput` and `digest::ExtendableOutput` traits, so it will be more consistent.

Reply:
Cool, will move these into the "core" traits